# ⚙️ แบบฝึกปฏิบัติการ (Workshop)  (ส่วนที่ 2/2) <br>
### การวิเคราะห์ข้อมูลและนำเสนอข้อมูลเบื้องต้น  <br>  COM104: Data Analytics and Visualization with Python 🐍 💻

---

## ℹ️ เกี่ยวกับ Notebook นี้

> Notebook นี้จัดทำขึ้นเพื่อประกอบการเรียนการสอนในรายวิชา COM104: การวิเคราะห์ข้อมูล และนำเสนอข้อมูลเบื้องต้นด้วย การเขียนโปรแกรมภาษาไพธอน (Data Analytics and Visualization with Python) 

🚀 ภายใต้โครงการอบรมเชิงปฏิบัติการ [Comcamp ครั้งที่ 34](https://www.facebook.com/KMUTTcomcamp) ระหว่างวันที่ 9-13 เมษายน 2566

⚙️ ณ ภาควิชาวิศวกรรมคอมพิวเตอร์ คณะวิศวกรรมศาสตร์ มหาวิทยาลัยเทคโนโลยีพระจอมเกล้าธนบุรี


---

##เตรียมความพร้อมNotebook

In [1]:
#นำเข้า Library พื้นฐานที่จำเป็น 
import pandas as pd 
import numpy as np
import math

import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns

pd.set_option('display.float_format', lambda x: '%.2f' % x)

## 0. ทำความเข้าใจโจทย์ปัญหาก่อนเริ่มกิจกรรม

## 1. การนำเข้าข้อมูล

**🚀 เป้าหมาย:**

* นำเข้าชุดข้อมูลผ่าน URL มาจัดเก็บในตัวแปร `data` ในรูปแบบ DataFrame 

**🧐 คำแนะนำ:**

* URL ชุดข้อมูล: https://raw.githubusercontent.com/ChotanansubSoph/COMCAMP-COM104/main/Dataset/electronic_store_fomatted.csv
* ใช้คำสั่ง [`pd.read_csv()`](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) อ่านข้อมูลไฟล์ csv มาจัดเก็บในตัวแปร `data`
* แสดงผล DataFrame

**💡 Hint**
```Python
url = ...
data = pd.read_csv(...)
data

```

**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

|    | event_time              |            order_id |          product_id |   category_id | category_code               | brand   |   price |     user_id |
|---:|:------------------------|--------------------:|--------------------:|--------------:|:----------------------------|:--------|--------:|------------:|
|  0 | 2020-04-24 11:50:39 UTC | 2294359932054536986 | 1515966223509089906 |   2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  1 | 2020-04-24 11:50:39 UTC | 2294359932054536986 | 1515966223509089906 |   2268105426648171008	 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  2 | 2020-04-24 14:37:43 UTC | 2294444024058086220 | 2273948319057183658 |   2268105426648171008	 | electronics.audio.headphone | huawei  |   77.52 | 1515915625441993984 |
|  3 | 2020-04-24 14:37:43 UTC | 2294444024058086220 | 2273948319057183658 |   2268105426648171008	 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 |
|  4 | 2020-04-24 19:16:21 UTC | 2294584263154074236 | 2273948316817424439 |   2268105426648171008	 | nan                         | karcher |  217.57 | 1515915625443148032 |
|  ... | ... | ... | ... | ...| ...    | ... |   ... | ...| ...      |
| 199995 | 2020-01-17 12:55:57 UTC | 2348774893979435140 | 1515966223523303310 | nan           | 8.10                      | nan     |  nan    | nan           |
| 199996 | 2020-01-17 12:55:57 UTC | 2348774893979435140 | 1515966223509104165 |   2268105428166509056 | electronics.smartphone    | huawei  |  416.64 | nan           |
| 199997 | 2020-01-17 12:55:57 UTC | 2348774893979435140 | 1515966223527092043 | nan           | 4.63                      | nan     |  nan    | nan           |
| 199998 | 2020-01-17 12:55:57 UTC | 2348774893979435140 | 1515966223523303312 | nan           | 6.94                      | nan     |  nan    | nan           |
| 199999 | 2020-01-18 07:00:00 UTC | 2348774894214316165 | 2273948226270790619 |   2268105441856717568	 | appliances.kitchen.kettle | vitek   |   18.5  |   1515915625484625152 |

**🧑🏻‍💻 ทดลองเขียนโค้ด:**

In [2]:
#เขียนโค้ดของน้องๆ ตรงนี้ได้เลย
url = "https://raw.githubusercontent.com/ChotanansubSoph/COMCAMP-COM104/main/Dataset/electronic_store_fomatted.csv"
data = pd.read_csv(url)



#ตรวจสอบผลลัพธ์ 
data

event_time             order_id           product_id  \
0       2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
1       2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
2       2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
3       2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
4       2020-04-24 19:16:21 UTC  2294584263154074236  2273948316817424439   
...                         ...                  ...                  ...   
199995  2020-01-17 12:55:57 UTC  2348774893979435140  1515966223523303310   
199996  2020-01-17 12:55:57 UTC  2348774893979435140  1515966223509104165   
199997  2020-01-17 12:55:57 UTC  2348774893979435140  1515966223527092043   
199998  2020-01-17 12:55:57 UTC  2348774893979435140  1515966223523303312   
199999  2020-01-18 07:00:00 UTC  2348774894214316165  2273948226270790619   

                  category_id                category_code    brand  price  \
0      2268105426648171008.00           electronics.tablet  samsung 162.01   
1      2268105426648171008.00           electronics.tablet  samsung 162.01   
2      2268105430162997504.00  electronics.audio.headphone   huawei  77.52   
3      2268105430162997504.00  electronics.audio.headphone   huawei  77.52   
4      2268105471367840000.00                          NaN  karcher 217.57   
...                       ...                          ...      ...    ...   
199995                    NaN                         8.10      NaN    NaN   
199996 2268105428166509056.00       electronics.smartphone   huawei 416.64   
199997                    NaN                         4.63      NaN    NaN   
199998                    NaN                         6.94      NaN    NaN   
199999 2268105441856717568.00    appliances.kitchen.kettle    vitek  18.50   

                      user_id  
0      1515915625441993984.00  
1      1515915625441993984.00  
2      1515915625447879424.00  
3      1515915625447879424.00  
4      1515915625443148032.00  
...                       ...  
199995                    NaN  
199996                    NaN  
199997                    NaN  
199998                    NaN  
199999 1515915625484625152.00  

[200000 rows x 8 columns]

##2. การตรวจสอบข้อมูลเบื้องต้น

###2.1 ตรวจสอบประเภทข้อมูล

**🚀 เป้าหมาย:**

* ตรวจสอบประเภทและจำนวนของข้อมูล


**🧐 คำแนะนำ:**
* ใช้คำสั่ง [`.info()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html) ในการตรวจสอบข้อมูล 


**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   event_time     50000 non-null  object 
 1   order_id       50000 non-null  int64  
 2   product_id     50000 non-null  int64  
 3   category_id    47090 non-null  float64
 4   category_code  38784 non-null  object 
 5   brand          45066 non-null  object 
 6   price          47090 non-null  float64
 7   user_id        47090 non-null  float64
dtypes: float64(3), int64(2), object(3)
memory usage: 3.1+ MB
```

In [ ]:
#เขียนโค้ดของน้องๆ ตรงนี้ได้เลย


##3. การทำความสะอาดข้อมูล

### 3.1 การปรับประเภทข้อมูลให้เหมาะสม

**🚀 เป้าหมาย:**

* ปรับประเภทข้อมูลที่เก็บรหัสตัวเลขต่างๆ จากเดิม int หรือ float ให้เป็น string


**💡 Hint**

```python
#ปรับทั้งสิ้น 4 คอลัมม์
data[ชื่อคอลัมม์เจ้าปัญหาที่ 1] = data['order_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 2] = data['product_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 3] = data['category_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 4] = data['user_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)

data.info()
```

**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

สังเกตคอลัมป์ `Dtype` และลองเปรียบเทียบกับผลลัพธ์การแสดงผลในข้อ 2

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   event_time     50000 non-null  object 
 1   order_id       50000 non-null  object 
 2   product_id     50000 non-null  object 
 3   category_id    50000 non-null  object 
 4   category_code  38784 non-null  object 
 5   brand          45066 non-null  object 
 6   price          47090 non-null  float64
 7   user_id        50000 non-null  object 
dtypes: float64(1), object(7)
memory usage: 3.1+ MB
```

[link text](https:// [link text](https://))**🧑🏻‍💻 ทดลองเขียนโค้ด:**

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้ได้เลย
data[ชื่อคอลัมม์เจ้าปัญหาที่ 1] = data['order_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 2] = data['product_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 3] = data['category_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)
data[ชื่อคอลัมม์เจ้าปัญหาที่ 4] = data['user_id'].apply(lambda x: '{:.0f}'.format(x)).astype(str)


SyntaxError: ignored

###3.2 การลบแถวที่ข้อมูลไม่ครบถ้วน

**🚀 เป้าหมาย:**

* ลบแถวที่มีข้อมูลไม่ครบถ้วน


**🧐 คำแนะนำ:**
* ลบข้อมูลโดยพิจารณาจาก ข้อความกำกับ `NaN` 
* สามารถทำได้ผ่านคำสั่ง [`.dropna()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html)
* รีเซ็ต Index ผ่านคำสั่ง [`reset_index()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)
 * ในที่นี้ใช้ `reset_index(drop=True)`
* ตรวจสอบประเภทข้อมูลอีกครั้งผ่านคำสั่ง [`.info()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html) 

**💡 Hint**

```python
data = ... #เรียกใช้ .dropna()
data = ... #เรียกใช้ .reset_index(drop=True)
```


**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

|    | event_time              |            order_id |          product_id |         category_id | category_code               | brand   |   price |             user_id |
|---:|:------------------------|--------------------:|--------------------:|--------------------:|:----------------------------|:--------|--------:|--------------------:|
|  0 | 2020-04-24 11:50:39 UTC | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  1 | 2020-04-24 11:50:39 UTC | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  2 | 2020-04-24 14:37:43 UTC | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 |
|  3 | 2020-04-24 14:37:43 UTC | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 |
|  4 | 2020-04-26 08:45:57 UTC | 2295716521449619456 | 1515966223509261824 | 2268105442636858112 | furniture.kitchen.table     | maestro |   39.33 | 1515915625450382848 |
|  ... | ... | ... | ... | ...| ...    | ... |   ... | ...| ...      |
| 111156 | 2020-01-18 09:28:02 UTC | 2348774892704366592 | 1515966223509118464 | 2268105428074234368 | electronics.telephone            | nokia    |   22.89 | nan           |
| 111157 | 2020-01-17 10:17:15 UTC | 2348774892737921024 | 1515966223509104896 | 2268105441856717568 | appliances.kitchen.kettle        | ava      |   13.87 | nan           |
| 111158 | 2020-01-17 06:18:58 UTC | 2348774892939247616 | 1515966223510058496 | 2268105393848713984 | appliances.kitchen.refrigerators | liebherr | 1226.83 | nan           |
| 111159 | 2020-01-17 12:55:57 UTC | 2348774893979435008 | 1515966223509104128 | 2268105428166509056 | electronics.smartphone           | huawei   |  416.64 | nan           |
| 111160 | 2020-01-18 07:00:00 UTC | 2348774894214316032 | 2273948226270790656 | 2268105441856717568 | appliances.kitchen.kettle        | vitek    |   18.5  |   1.51592e+18 |

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้ได้เลย!


#ทดสอบแสดงผล
data

###3.3 การปรับรูปแบบข้อมูลเวลา

**🚀 เป้าหมาย:**

* ปรับข้อมูลเวลา ให้อยู่ในรูปแบบข้อเวลาตามมาตรฐานของ Pandas


**💡 Hint**

```python
data[ชื่อคอลัมม์เวลา] = pd.to_datetime(data[ชื่อคอลัมม์เวลา])
```


**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

|    | event_time                |            order_id |          product_id |         category_id | category_code               | brand   |   price |             user_id |
|---:|:--------------------------|--------------------:|--------------------:|--------------------:|:----------------------------|:--------|--------:|--------------------:|
|  0 | 2020-04-24 11:50:39+00:00 | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  1 | 2020-04-24 11:50:39+00:00 | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 |
|  2 | 2020-04-24 14:37:43+00:00 | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 |
|  3 | 2020-04-24 14:37:43+00:00 | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 |
|  4 | 2020-04-26 08:45:57+00:00 | 2295716521449619456 | 1515966223509261824 | 2268105442636858112 | furniture.kitchen.table     | maestro |   39.33 | 1515915625450382848 |
|  ... | ... | ... | ... | ...| ...    | ... |   ... | ...| ...      |
| 111156 | 2020-01-18 09:28:02+00:00 | 2348774892704366592 | 1515966223509118464 | 2268105428074234368 | electronics.telephone            | nokia    |   22.89 | nan           |
| 111157 | 2020-01-17 10:17:15+00:00 | 2348774892737921024 | 1515966223509104896 | 2268105441856717568 | appliances.kitchen.kettle        | ava      |   13.87 | nan           |
| 111158 | 2020-01-17 06:18:58+00:00 | 2348774892939247616 | 1515966223510058496 | 2268105393848713984 | appliances.kitchen.refrigerators | liebherr | 1226.83 | nan           |
| 111159 | 2020-01-17 12:55:57+00:00 | 2348774893979435008 | 1515966223509104128 | 2268105428166509056 | electronics.smartphone           | huawei   |  416.64 | nan           |
| 111160 | 2020-01-18 07:00:00+00:00 | 2348774894214316032 | 2273948226270790656 | 2268105441856717568 | appliances.kitchen.kettle        | vitek    |   18.5  |   1.51592e+18 |

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้!!


#การแสดงผลข้อมูล
data

###3.4 เพิ่มคอลัมม์ประเภทข้อมูลสินค้า

**🚀 เป้าหมาย:**

* ดึงข้อมูลประเภทสินค้า ออกมาจากรหัสประเภทข้อมูลสินค้า
  * เช่น จากเดิม electronics.tablet เป็น tablet เป็นต้น


**💡 Hint** 

คำใบ้ที่ 1
```python
def get_catagory(code):
  return code.split(ตัวคั่นที่ต้องการแยก)[-1]
```

คำใบ้ที่ 2 
```python
data[คอลัมป์ที่สนใจ] = data[คอลัมป์ที่สนใจ].apply(ฟังก์ชัน)
```
**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

|    | event_time                |            order_id |          product_id |         category_id | category_code               | brand   |   price |             user_id | category   |
|---:|:--------------------------|--------------------:|--------------------:|--------------------:|:----------------------------|:--------|--------:|--------------------:|:-----------|
|  0 | 2020-04-24 11:50:39+00:00 | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 | tablet     |
|  1 | 2020-04-24 11:50:39+00:00 | 2294359932054536960 | 1515966223509089792 | 2268105426648171008 | electronics.tablet          | samsung |  162.01 | 1515915625441993984 | tablet     |
|  2 | 2020-04-24 14:37:43+00:00 | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 | headphone  |
|  3 | 2020-04-24 14:37:43+00:00 | 2294444024058086144 | 2273948319057183744 | 2268105430162997504 | electronics.audio.headphone | huawei  |   77.52 | 1515915625447879424 | headphone  |
|  4 | 2020-04-26 08:45:57+00:00 | 2295716521449619456 | 1515966223509261824 | 2268105442636858112 | furniture.kitchen.table     | maestro |   39.33 | 1515915625450382848 | table      |
|  ... | ... | ... | ... | ...| ...    | ... |   ... | ...| ...      |
| 111156 | 2020-01-18 09:28:02+00:00 | 2348774892704366592 | 1515966223509118464 | 2268105428074234368 | electronics.telephone            | nokia    |   22.89 | nan           | telephone     |
| 111157 | 2020-01-17 10:17:15+00:00 | 2348774892737921024 | 1515966223509104896 | 2268105441856717568 | appliances.kitchen.kettle        | ava      |   13.87 | nan           | kettle        |
| 111158 | 2020-01-17 06:18:58+00:00 | 2348774892939247616 | 1515966223510058496 | 2268105393848713984 | appliances.kitchen.refrigerators | liebherr | 1226.83 | nan           | refrigerators |
| 111159 | 2020-01-17 12:55:57+00:00 | 2348774893979435008 | 1515966223509104128 | 2268105428166509056 | electronics.smartphone           | huawei   |  416.64 | nan           | smartphone    |
| 111160 | 2020-01-18 07:00:00+00:00 | 2348774894214316032 | 2273948226270790656 | 2268105441856717568 | appliances.kitchen.kettle        | vitek    |   18.5  |   1.51592e+18 | kettle        |

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้ได้เลย


#ตรวจสอบ
data

## 4.การวิเคราะห์และนำเสนอข้อมูลเบื้องต้น ผ่านโจทย์ปัญหาต่างๆ

### 4.1 แบรนด์สินค้าไหนที่ทำรายได้ดี?

**🚀 เป้าหมาย:**

* นำเสนอข้อมูล 10 อันดับแบรนด์สินค้าที่ทำรายได้สูงสุด 


**🧐 คำแนะนำ:**

* ใช้ `.groupby()` ในการจัดกลุ่มข้อมูล
* หาผลรวมข้อมูลในแต่ละกลุ่มโดยใช้ `.sum()`
* รีเซ็ต index โดยใช้ `reset_idex()`
* เรียงลำดับจากมากไปน้อยโดยใช้ `.sort_value()`
 * ในที่นี้ใช้ `.sort_value(ชื่อคอลัมม์ที่ต้องการใช้เป็นเกณฑ์, ascending=False)`
* นำเสนอข้อมูลผ่าน `px.bar`

**💡 Hint** 

**A แทน ชื่อคอลัมป์ที่ต้องการจัดลุ่ม (แบรนด์)

**B แทน ชื่อคอลัมป์ที่ต้องการใช้ในการประกอบการจำแนก (ราคา)

```python
best_performing_brands = data.groupby(**A)[**B].sum()
best_performing_brands = best_performing_brands.reset_index().sort_values(**B, ascending=False).head(10)

fig = px.bar(
    best_performing_brands, 
    x='brand', 
    y='price', 
    title='Best performing brands',
    width=800, 
    height=800
)

```



**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

![](https://drive.google.com/uc?id=1B7vv_DfLV8c5ez43LndXV2dxa642xWRM)

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้!

#แสดงผล
fig.show()

### 4.2 แบรนด์สินค้าไหนขายดี?

**🚀 เป้าหมาย:**

*  นำเสนอแผนภูมิแท่งแสดงข้อมูล 10 อันดับแบรนด์สินค้าที่ขายดีที่สุด (พิจารณาจากจำนวนสินค้าที่ขายได้)


**🧐 คำแนะนำ:**

* ใช้ `.agg('count')` ในการสรุปปริมาณ


**💡 Hint** 

**A แทน ชื่อคอลัมป์ที่ต้องการจัดลุ่ม (แบรนด์)

**B แทน ชื่อคอลัมป์ที่ต้องการใช้ในการประกอบการจำแนก (ราคา)

```python
most_sold_brands = data.groupby(**A)[**B].agg('count').reset_index().sort_values('price', ascending=False).head(10)
most_sold_brands.rename(columns={**A: **A, **B: "times_sold"}, inplace=True)

fig = px.bar(...) #คล้ายคลึงกับข้อ 4.1 ข้อมูลใช้เป็น most_sold_brands เปลี่ยน x='brand', y='time_sold" และ title ตั้งชื่อตามความเหมาะสม

```


**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

![](https://drive.google.com/uc?id=1_D8o2zhQqpqWf4IXYvwnzI_2skp9LCs_)

In [ ]:


fig = px.bar(
    most_sold_brands, 
    x='brand', 
    y='times_sold', 
    title='Most sold brands',
    width=800, 
    height=800
)

fig.show()

### 4.3 ประเภทสินค้าอะไรที่ขายดี?

**🚀 เป้าหมาย:**

* นำเสนอแผนภูมิแท่งแสดงข้อมูล 10 อันดับประเภทสินค้าที่ขายดีที่สุด (พิจารณาจากจำนวนสินค้าที่ขายได้)


**🧐 คำแนะนำ:**



**💡 Hint** 



```python
#คล้ายคลึงกับข้อก่อนหน้า โดยเปลี่ยน **A เป็น ประเภทสินค้า (category)
most_sold_brands = data.groupby(**A)[**B].agg('count').reset_index().sort_values('price', ascending=False).head(10)
most_sold_brands.rename(columns={**A: **A, **B: "times_sold"}, inplace=True)

fig = px.ber(...) #ปรับเปลี่ยน title ตามความเหมาะสม

```

**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

![](https://drive.google.com/uc?id=1nxLdudUh_2lcdLC5_kqOHu35-cFw-wGf)

In [ ]:
#เขียนโค้ดของน้องๆ ตรงนี้!!!

#ส่วนแสดงผล
fig.show()

### 4.4 ประเภทสินค้าอะไรที่ทำรายได้ดี? [BONUS🌟]

**🚀 เป้าหมาย:**

* แสดงแผนภูมิวงกลมแสดงสัดส่วนของรายได้สินค้า




**💡 Hint** 

**A แทน ชื่อคอลัมป์ที่ต้องการจัดลุ่ม (ประเภทสินค้า) 

**B แทน ชื่อคอลัมป์ที่ต้องการใช้ในการประกอบการจำแนก (ราคา)

```python
grouped = data.groupby(**A)[**B].sum().reset_index()
sorted_grouped = grouped.sort_values(**B, ascending=False)[:9]

other_price = grouped.loc[~grouped[**A].isin(sorted_grouped[**A].values), **B].sum()
other = pd.DataFrame({**A: ['Other'], **B: [other_price]})
top_10 = pd.concat([sorted_grouped, other])

fig = px.pie(top_10, values=**B, names=**A, title="")
fig.show()
```

**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

![](https://drive.google.com/uc?id=1hM7byH54KR0h47nxgPwtyhGGUVKfkdN8)

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้ได้เลย

#ส่วนแสดงผล
fig.show()

### 4.5 สินค้าถูกซื้อช่วงเวลาไหนบ้างนะ? [BONUS🌟]

**🚀 เป้าหมาย:**

* นำเสนอแผนภูมิแท่งแสดงช่วงเวลาที่สินค้าถูกซื้อมากที่สุด



**💡 Hint** 

**A แทนชื่อคอลัมม์ที่เก็บข้อมูลที่ต้องการจัดกลุ่ม (เวลา)

```python
best_time = data.groupby(data[**A].dt.strftime('%r'))['price'].sum().sort_values()[lambda x: x<= 900000].tail(10)

fig = px.bar(
    best_time,
    x="price", 
    orientation='h', 
    title=" ", #ตั้งชื่อตามความเหมาะสม 
    width=800, 
    height=800
)
```

**✅ ตัวอย่างผลลัพธ์การแสดงผล:**

![](https://drive.google.com/uc?id=16O36hvKv9OVE4V6nDYDWFWvdMV_LsGk8)

In [ ]:
#เขียนโค้ดของน้องๆตรงนี้!


#ส่วนแสดงผล
fig.show()

## 📚 อ้างอิง (Reference)


---
> **📊 ชุดข้อมูลที่ใช้ในกิจกรรม**

ได้หยิบยกข้อมูลมาจากแหล่งข้อมูลออนไลน์ ดังแสดงในรายละเอียดต่อไปนี้

* ชื่อชุดข้อมูล: eCommerce purchase history from electronics store

* ผู้เผยแพร่: MICHAEL KECHINOV

* แหล่งที่มา: https://www.kaggle.com/datasets/mkechinov/ecommerce-purchase-history-from-electronics-store

---

> **💡 แนวทางการวิเคราะห์ข้อมูลในกิจกรรม** 

ได้หยิบยกข้อมูลมาจากแหล่งข้อมูลออนไลน์ ดังแสดงในรายละเอียดต่อไปนี้
* หัวข้อ: Purchase Analysis

* ผู้เผยแพร่: TANVIR SIDDIKE MOIN

* แหล่งที่มา: https://www.kaggle.com/code/tanvirsiddikemoin/purchase-analysis